File: Code/Avernakis Nodejs/Test-Nodejs/src/Ave/Ui/Control/UiComboBox.ts
import { IControl } from "../UiControl";
import { AveLib } from "../../AveLib";
import { DpiSize, StringKey } from "../UiCommon";
import { Window } from "./UiWindow";

export enum ComboBoxMode {
    DropDown,
    DropDownEdit,
    ListEdit,
}

export interface IComboBox extends IControl {
    new (window: Window, key?: StringKey): IComboBox;

    Append(s: string): ComboBox;
    Insert(s: string, nInsertBefore: number): ComboBox;
    Remove(nIndex: number): boolean;
    Clear(): ComboBox;
    Select(nIndex: number): ComboBox;
    Sort(bAsc: boolean, bCaseInsensitive: boolean): ComboBox;
    Find(
        s: string,
        nIndexStart: number,
        bExact: boolean,
        bCaseInsensitive: boolean
    ): number;
    Set(nIndex: number, s: string): boolean;
    Get(nIndex: number): string;
    GetCount(): number;
    GetSelection(): number;

    SetCount(nCount: number): ComboBox;

    SetMode(nMode: ComboBoxMode): ComboBox;
    GetMode(): ComboBoxMode;

    SetDropHeight(n: DpiSize): ComboBox;
    GetDropHeight(): DpiSize;

    SetFindExact(b: boolean): ComboBox;
    GetFindExact(): boolean;

    SetFindCaseInsensitive(b: boolean): ComboBox;
    GetFindCaseInsensitive(): boolean;

    SetBorder(b: boolean): ComboBox;
    GetBorder(): boolean;

    OnSelectionChange(fn: (sender: ComboBox) => void): ComboBox;
}

export class ComboBox extends (AveLib.UiComboBox as IComboBox) {
    Append(...s: string[]) {
        s.forEach((e) => super.Append(e));
        return this;
    }
}

File: Code/Avernakis Nodejs/Avernakis-Nodejs/Private/UiComboBox.cpp
#include "StdAfx.h"
#include "UiComboBox.h"

#define ThisMethod($x) &UiComboBox::$x
#define AutoAddMethod($x, ...) AddMethod<__VA_ARGS__>( #$x, ThisMethod( $x ) )
#define MakeThisFunc($x) MakeFunc( this, &UiComboBox::$x )

namespace Nav
{

	namespace
	{
		ObjectRegister<UiComboBox> c_obj;
	}

	void UiComboBox::DefineControl()
	{
		AutoAddMethod( Append );
		AutoAddMethod( Insert );
		AutoAddMethod( Remove );
		AutoAddMethod( Clear );
		AutoAddMethod( Select );
		AutoAddMethod( Sort );
		AutoAddMethod( Find, WrapObjectGeneric );
		AutoAddMethod( Set );
		AutoAddMethod( Get, WrapObjectGeneric );
		AutoAddMethod( GetCount, WrapObjectGeneric );
		AutoAddMethod( GetSelection, WrapObjectGeneric );

		AutoAddMethod( SetCount );

		AutoAddMethod( SetMode );
		AutoAddMethod( GetMode, WrapObjectGeneric );

		AutoAddMethod( SetDropHeight );
		AutoAddMethod( GetDropHeight, WrapObjectGeneric );

		AutoAddMethod( SetFindExact );
		AutoAddMethod( GetFindExact, WrapObjectGeneric );

		AutoAddMethod( SetFindCaseInsensitive );
		AutoAddMethod( GetFindCaseInsensitive, WrapObjectGeneric );

		AutoAddMethod( SetBorder );
		AutoAddMethod( GetBorder, WrapObjectGeneric );

		AutoAddMethod( OnSelectionChange, WrapObjectGeneric );
	}

	U1 UiComboBox::Ctor( UiWindow * p, Napi::Value v )
	{
		if ( !__CreateControl( p, v ) )
			return false;

		return true;
	}

	void UiComboBox::__OnSelectionChange( Ui::IComboBox & sender )
	{
		m_OnSelectionChange( this );
	}

	UiComboBox * UiComboBox::OnSelectionChange( OnSelectionChange_t && fn )
	{
		m_OnSelectionChange = SetEventCallback<Ui::IComboBox::OnSelectionChange>( std::move( fn ), MakeThisFunc( __OnSelectionChange ) );
		return this;
	}

}
